hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+                }
+
+                return conn;
+            }
+
+
+            std::vector<boost::shared_ptr<Connection> > ClientConnectionManagerImpl::getActiveConnections() {
+                return activeConnections.values();
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::getOrTriggerConnect(const Address &target) {
+                boost::shared_ptr<Connection> connection = getConnection(target, false);
+                if (connection.get() != NULL) {
+                    return connection;
+                }
+                triggerConnect(target, false);
+                return boost::shared_ptr<Connection>();
+            }
+
+            boost::shared_ptr<Connection>
+            ClientConnectionManagerImpl::getConnection(const Address &target, bool asOwner) {
+                if (!asOwner) {
+                    connectionStrategy->beforeGetConnection(target);
+                }
+                if (!asOwner && getOwnerConnection().get() == NULL) {
+                    throw exception::IOException("ConnectionManager::getConnection",
+                                                 "Owner connection is not available!");
+                }
+
+                boost::shared_ptr<Connection> connection = activeConnections.get(target);
+
+                if (connection.get() != NULL) {
+                    if (!asOwner) {
+                        return connection;
+                    }
+                    if (connection->isAuthenticatedAsOwner()) {
+                        return connection;
+                    }
+                }
+                return boost::shared_ptr<Connection>();
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::getActiveConnection(const Address &target) {
+                return activeConnections.get(target);
+            }
+
+            boost::shared_ptr<Address> ClientConnectionManagerImpl::getOwnerConnectionAddress() {
+                return ownerConnectionAddress;
+            }
+
+            boost::shared_ptr<AuthenticationFuture>
+            ClientConnectionManagerImpl::triggerConnect(const Address &target, bool asOwner) {
+                if (!asOwner) {
+                    connectionStrategy->beforeOpenConnection(target);
+                }
+                if (!alive) {
+                    throw exception::HazelcastException("ConnectionManager::triggerConnect",
+                                                        "ConnectionManager is not active!");
+                }
+
+                boost::shared_ptr<AuthenticationFuture> future(new AuthenticationFuture());
+                boost::shared_ptr<AuthenticationFuture> oldFuture = connectionsInProgress.putIfAbsent(target, future);
+                if (oldFuture.get() == NULL) {
+                    executionService.execute(
+                            boost::shared_ptr<util::Runnable>(new InitConnectionTask(target, asOwner, future, *this)));
```

Comment:
Passing conection manager to an async  task could be dangerrous. 
Is there a way to make sure connection manager is not  deleted until this task is  finished ?

Reply:
We actually make sure that all threads exit before the client instance destructor exits where the connection manager and other spi services are being destroyed.